CodeGen Error in iOS build & druntime unit tests

Basically, this is from running a unit test build for iOS through my device. The problem is in 4/5 where `this.gcx` should become `this` when calling `gcx.alloc`, but `this*` is `NULL` after that call

```

Thread 1 Queue : com.apple.main-thread (serial)
#0	0x0000000104cf7700 in _D2gc9pooltable__T9PoolTableTSQBc4impl12conservativeQBy4PoolZQBr6lengthMxFNaNbNdNiNfZm at /opt/ios-testing/runtime/druntime/src/gc/pooltable.d:56
#1	0x0000000104cead68 in _D2gc4impl12conservativeQw3Gcx6npoolsMxFNaNbNdZm at /opt/ios-testing/runtime/druntime/src/gc/impl/conservative/gc.d:1236
#2	0x0000000104ced028 in _D2gc4impl12conservativeQw3Gcx8bigAllocMFNbmKmkxC8TypeInfoZ8tryAllocMFNbZb at /opt/ios-testing/runtime/druntime/src/gc/impl/conservative/gc.d:1721
#3	0x0000000104cec844 in _D2gc4impl12conservativeQw3Gcx8bigAllocMFNbmKmkxC8TypeInfoZPv at /opt/ios-testing/runtime/druntime/src/gc/impl/conservative/gc.d:1743
#4	0x0000000104ce700c in _D2gc4impl12conservativeQw3Gcx5allocMFNbmKmkxC8TypeInfoZPv at /opt/ios-testing/runtime/druntime/src/gc/impl/conservative/gc.d:1628

=> this	Gcx *	NULL	

#5	0x0000000104ce6f0c in _D2gc4impl12conservativeQw14ConservativeGC12mallocNoSyncMFNbmkKmxC8TypeInfoZPv at /opt/ios-testing/runtime/druntime/src/gc/impl/conservative/gc.d:389

=> this	ConservativeGC *	0x16cebf108	0x000000016cebf108
=>=> gcx	Gcx *	0x26cebf150	0x000000026cebf150

#6	0x0000000104ce6e40 in _D2gc4impl12conservativeQw14ConservativeGC__T9runLockedS_DQCeQCeQCcQCnQBs12mallocNoSyncMFNbmkKmxC8TypeInfoZPvS_DQEgQEgQEeQEp10mallocTimelS_DQFiQFiQFgQFr10numMallocslTmTkTmTxQCzZQFcMFNbKmKkKmKxQDsZQDl at /opt/ios-testing/runtime/druntime/src/gc/impl/conservative/gc.d:254
#7	0x0000000104ce70f8 in _D2gc4impl12conservativeQw14ConservativeGC6qallocMFNbmkxC8TypeInfoZS4core6memory8BlkInfo_ at /opt/ios-testing/runtime/druntime/src/gc/impl/conservative/gc.d:417
#8	0x0000000104ceaeac in _DThn16_2gc4impl12conservativeQw14ConservativeGC6qallocMFNbmkxC8TypeInfoZS4core6memory8BlkInfo_ at /opt/ios-testing/runtime/druntime/src/gc/impl/conservative/gc.d:407
#9	0x0000000104cf8340 in gc_qalloc at /opt/ios-testing/runtime/druntime/src/gc/proxy.d:176
#10	0x0000000104cf6960 in _D2gc4impl5protoQo7ProtoGC6qallocMFNbmkxC8TypeInfoZS4core6memory8BlkInfo_ at /opt/ios-testing/runtime/druntime/src/gc/impl/proto/gc.d:111
...
#17	0x0000000102f449ac in _D11test_runner24_sharedStaticCtor_L80_C1FZv at /opt/ios-testing/runtime/druntime/src/test_runner.d:92
...
#23	0x0000000104d0bb1c in rt_init at /opt/ios-testing/runtime/druntime/src/rt/dmain2.d:212

```

Not sure if it helps, but I added printfs to gc.d and I'm getting this error now when compiling:

```
FAILED: druntime-test-runner-shared.app/druntime-test-runner-shared 
: && /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/cc -DHAVE_UNISTD_H -arch arm64 -isysroot /Applications/Xcode.app/Contents/Developer/Platforms/iPhoneOS.platform/Developer/SDKs/iPhoneOS13.2.sdk -miphoneos-version-min=13.2 -Wl,-headerpad_max_install_names  objects-unittest-shared/test_runner.o  -o druntime-test-runner-shared.app/druntime-test-runner-shared  -Wl,-rpath,/Users/etcimon/ldc-build-runtime.tmp/lib /Users/etcimon/ldc-build-runtime.tmp/lib/libdruntime-ldc-unittest-shared.dylib && :
Undefined symbols for architecture arm64:
  "__D12TypeInfo_Aya6__initZ", referenced from:
      __D11test_runner24_sharedStaticCtor_L80_C1FZv in test_runner.o
  "__D4core4time13convClockFreqFNaNbNiNflllZl", referenced from:
      __D11test_runner6doTestFPS6object10ModuleInfoKS4core7runtime14UnitTestResultZv in test_runner.o
      __D4core4time__T12MonoTimeImplVEQBdQBb9ClockTypei0ZQBj__T8opBinaryVAyaa1_2dZQtMxFNaNbNiNfSQDjQDh__TQDfVQCui0ZQDpZSQEhQEf8Duration in test_runner.o
  "__D4core4time8Duration6__ctorMFNaNbNcNiNflZSQBpQBnQBl", referenced from:
      __D11test_runner6doTestFPS6object10ModuleInfoKS4core7runtime14UnitTestResultZv in test_runner.o
      __D4core4time__T12MonoTimeImplVEQBdQBb9ClockTypei0ZQBj__T8opBinaryVAyaa1_2dZQtMxFNaNbNiNfSQDjQDh__TQDfVQCui0ZQDpZSQEhQEf8Duration in test_runner.o
  "__D4core4time__T12MonoTimeImplVEQBdQBb9ClockTypei0ZQBj14ticksPerSecondFNaNbNdNiNfZl", referenced from:
      __D11test_runner6doTestFPS6object10ModuleInfoKS4core7runtime14UnitTestResultZv in test_runner.o
      __D4core4time__T12MonoTimeImplVEQBdQBb9ClockTypei0ZQBj__T8opBinaryVAyaa1_2dZQtMxFNaNbNiNfSQDjQDh__TQDfVQCui0ZQDpZSQEhQEf8Duration in test_runner.o
  "__D4core4time__T12MonoTimeImplVEQBdQBb9ClockTypei0ZQBj8currTimeFNbNdNiNeZSQCtQCr__TQCpVQCei0ZQCz", referenced from:
      __D11test_runner6doTestFPS6object10ModuleInfoKS4core7runtime14UnitTestResultZv in test_runner.o
  "__D4core7runtime7Runtime24extendedModuleUnitTesterFNdPFZSQCcQCa14UnitTestResultZv", referenced from:
      __D11test_runner24_sharedStaticCtor_L80_C1FZv in test_runner.o
  "__D4core7runtime7Runtime4argsFNdZAAya", referenced from:
      __D11test_runner6testerFZS4core7runtime14UnitTestResult in test_runner.o
      __D11test_runner11testModulesFZS4core7runtime14UnitTestResult in test_runner.o
  "__D4core8internal5array8equality__T8__equalsTyaTyaZQqFNaNbNiNfAyaQdZb", referenced from:
      __D11test_runner13getModuleInfoFAyaZ14__foreachbody2MFPS6object10ModuleInfoZi in test_runner.o
      __D11test_runner6testerFZS4core7runtime14UnitTestResult in test_runner.o
      __D11test_runner11testModulesFZS4core7runtime14UnitTestResult in test_runner.o
  "__D6object10ModuleInfo4nameMxFNaNbNdNiZAya", referenced from:
      __D11test_runner13getModuleInfoFAyaZ14__foreachbody2MFPS6object10ModuleInfoZi in test_runner.o
      __D11test_runner6doTestFPS6object10ModuleInfoKS4core7runtime14UnitTestResultZv in test_runner.o
  "__D6object10ModuleInfo7opApplyFMDFPSQBhQBdZiZi", referenced from:
      __D11test_runner13getModuleInfoFAyaZPS6object10ModuleInfo in test_runner.o
      __D11test_runner6testerFZS4core7runtime14UnitTestResult in test_runner.o
      __D11test_runner11testModulesFZS4core7runtime14UnitTestResult in test_runner.o
      __D11test_runner7testAllFZS4core7runtime14UnitTestResult in test_runner.o
  "__D6object10ModuleInfo8unitTestMxFNaNbNdNiZPFZv", referenced from:
      __D11test_runner6doTestFPS6object10ModuleInfoKS4core7runtime14UnitTestResultZv in test_runner.o
  "__D6object9Throwable7__ClassZ", referenced from:
      GCC_except_table5 in test_runner.o
  "__d_arrayappendT", referenced from:
      __D11test_runner24_sharedStaticCtor_L80_C1FZv in test_runner.o
  "__d_dso_registry", referenced from:
      _ldc.register_dso in test_runner.o
  "__d_eh_enter_catch", referenced from:
      __D11test_runner6doTestFPS6object10ModuleInfoKS4core7runtime14UnitTestResultZv in test_runner.o
  "__d_eh_personality", referenced from:
      Dwarf Exception Unwind Info (__eh_frame) in test_runner.o
  "__d_run_main", referenced from:
      _main in test_runner.o
ld: symbol(s) not found for architecture arm64
```

The patched LDC is being used that's for sure, the linking fails so I do it through xcode with the same build for which the debug test runs work. On a separate note dub fails with `building for macOS-x86_64 but attempting to link with file built for iOS-arm64`

```
Configuring dependent libasync, deps:"memutils"
  Configuring dependent memutils, deps:
Performing "unittest" build using ldc2 for aarch64, arm_hardfloat.
memutils 1.0.4: target for configuration "secure" is up to date.
Using existing build in /var/root/.dub/packages/memutils-1.0.4/memutils/.dub/build/secure-unittest-posix-aarch64.arm_hardfloat-ldc_2090-21DB8A81048C577DAFEBE2E0B90DDB53/.
Copying target from /var/root/.dub/packages/memutils-1.0.4/memutils/.dub/build/secure-unittest-posix-aarch64.arm_hardfloat-ldc_2090-21DB8A81048C577DAFEBE2E0B90DDB53/libmemutils.a to /var/root/.dub/packages/memutils-1.0.4/memutils
Target '/opt/libasync/.dub/build/libasync-test-regular-unittest-posix-aarch64.arm_hardfloat-ldc_2090-6124DC0CD5CE6FBFFDA970EA7C91A6E7/libasync-test-regular' doesn't exist, need rebuild.
libasync 0.8.3+commit.27.gf976020: building configuration "libasync-test-regular"...
ldc2 -mtriple=aarch64-apple-ios13.2 -c -of.dub/build/libasync-test-regular-unittest-posix-aarch64.arm_hardfloat-ldc_2090-6124DC0CD5CE6FBFFDA970EA7C91A6E7/libasync-test-regular.o -d-debug -g -unittest -w -oq -od=.dub/obj -d-version=EnableDebugger -d-version=Have_libasync -d-version=Have_memutils -d-version=CryptoSafe -Isource/ -I/var/root/.dub/packages/memutils-1.0.4/memutils/source/ /tmp/dub_test_root_a10414e4_1550_4f81_97ef_83024019611b.d source/libasync/bufferedtcp.d source/libasync/dns.d source/libasync/event.d source/libasync/events.d source/libasync/file.d source/libasync/internals/epoll.d source/libasync/internals/freelist.d source/libasync/internals/hashmap.d source/libasync/internals/kqueue.d source/libasync/internals/logging.d source/libasync/internals/memory.d source/libasync/internals/path.d source/libasync/internals/queue.d source/libasync/internals/socket_compat.d source/libasync/internals/validator.d source/libasync/internals/win32.d source/libasync/notifier.d source/libasync/package.d source/libasync/posix.d source/libasync/posix2.d source/libasync/signal.d source/libasync/socket.d source/libasync/tcp.d source/libasync/test.d source/libasync/threads.d source/libasync/timer.d source/libasync/types.d source/libasync/udp.d source/libasync/uds.d source/libasync/watcher.d source/libasync/windows.d -vcolumns
Enhanced memory security is enabled.
Memory debugger enabled
Using FreeBSD KQueue for events
source/libasync/test.d(57,11): Deprecation: struct std.datetime.StopWatch is deprecated - To be removed after 2.094. Use std.datetime.stopwatch.StopWatch instead.
source/libasync/test.d(57,11): Deprecation: struct std.datetime.StopWatch is deprecated - To be removed after 2.094. Use std.datetime.stopwatch.StopWatch instead.
Linking...
ldc2 -of.dub/build/libasync-test-regular-unittest-posix-aarch64.arm_hardfloat-ldc_2090-6124DC0CD5CE6FBFFDA970EA7C91A6E7/libasync-test-regular .dub/build/libasync-test-regular-unittest-posix-aarch64.arm_hardfloat-ldc_2090-6124DC0CD5CE6FBFFDA970EA7C91A6E7/libasync-test-regular.o /var/root/.dub/packages/memutils-1.0.4/memutils/.dub/build/secure-unittest-posix-aarch64.arm_hardfloat-ldc_2090-21DB8A81048C577DAFEBE2E0B90DDB53/libmemutils.a -mtriple=aarch64-apple-ios13.2 -g
ld: warning: ignoring file /var/root/.dub/packages/memutils-1.0.4/memutils/.dub/build/secure-unittest-posix-aarch64.arm_hardfloat-ldc_2090-21DB8A81048C577DAFEBE2E0B90DDB53/libmemutils.a, building for macOS-x86_64 but attempting to link with file built for iOS-arm64
ld: warning: ignoring file .dub/build/libasync-test-regular-unittest-posix-aarch64.arm_hardfloat-ldc_2090-6124DC0CD5CE6FBFFDA970EA7C91A6E7/libasync-test-regular.o, building for macOS-x86_64 but attempting to link with file built for iOS-arm64
ld: warning: ignoring file /opt/build-ldc-testing/ldc-build-runtime.tmp/lib/libdruntime-ldc.a, building for macOS-x86_64 but attempting to link with file built for iOS-arm64
ld: warning: ignoring file /opt/build-ldc-testing/ldc-build-runtime.tmp/lib/libphobos2-ldc.a, building for macOS-x86_64 but attempting to link with file built for iOS-arm64
Undefined symbols for architecture x86_64:
  "_main", referenced from:
     implicit entry/start for main executable
ld: symbol(s) not found for architecture x86_64
clang: error: linker command failed with exit code 1 (use -v to see invocation)
Error: /usr/bin/cc failed with status: 1
FAIL .dub/build/libasync-test-regular-unittest-posix-aarch64.arm_hardfloat-ldc_2090-6124DC0CD5CE6FBFFDA970EA7C91A6E7/ libasync-test-regular executable
ldc2 failed with exit code 1.
```

--- a/ir/irclass.cpp
+++ b/ir/irclass.cpp
@@ -203,7 +203,8 @@ LLConstant *IrAggr::getVtblInit() {
         Logger::println("Running late functionSemantic to infer return type.");
         if (!fd->functionSemantic()) {
           if (fd->semantic3Errors) {
-            Logger::println("functionSemantic failed; using null for vtbl entry.");
+            Logger::println(
+                "functionSemantic failed; using null for vtbl entry.");
             constants.push_back(getNullValue(voidPtrType));
             continue;
           }
@@ -301,8 +302,8 @@ llvm::GlobalVariable *IrAggr::getInterfaceVtblSymbol(BaseClass *b,
 
   const auto irMangle = getIRMangledVarName(mangledName.peekChars(), LINKd);
 
-  LLGlobalVariable *gvar =
-      declareGlobal(cd->loc, gIR->module, vtblType, irMangle, /*isConstant=*/true);
+  LLGlobalVariable *gvar = declareGlobal(cd->loc, gIR->module, vtblType,
+                                         irMangle, /*isConstant=*/true);
 
   // insert into the vtbl map
   interfaceVtblMap.insert({{b->sym, interfaces_index}, gvar});
@@ -462,9 +463,10 @@ void IrAggr::defineInterfaceVtbl(BaseClass *b, bool new_instance,
 
       // call the real vtbl function.
       llvm::CallInst *call = gIR->ir->CreateCall(callee, args);
-      call->setCallingConv(irFunc->getCallingConv());
-      call->setTailCallKind(thunk->isVarArg() ? llvm::CallInst::TCK_MustTail
-                                              : llvm::CallInst::TCK_Tail);
+      call->setCallingConv(callee->getCallingConv());
+      call->setAttributes(callee->getAttributes());
+      call->setTailCallKind(callee->isVarArg() ? llvm::CallInst::TCK_MustTail
+                                               : llvm::CallInst::TCK_Tail);
 
       // return from the thunk
       if (thunk->getReturnType() == LLType::getVoidTy(gIR->context())) {

--- a/shippable.yml
+++ b/shippable.yml
@@ -83,14 +83,14 @@ build:
     # Run LDC D unittests
     - ctest --output-on-failure -R "ldc2-unittest"
     # Run LIT testsuite, ignore the errors
-    - PATH=$PWD/../llvm/bin:$PATH ctest -V -R "lit-tests" || true
+    - cd tests
+    - PATH="$PWD/../../llvm/bin:$PATH" python runlit.py -v -j 32 . || true
+    - cd ..
     # Run DMD testsuite, ignore the errors
-    # FIXME: the debug libs seem not really usable, most runnable tests crash
-    - sed -i 's|REQUIRED_ARGS=-g -link-defaultlib-debug|REQUIRED_ARGS=-g|g' tests/d2/CTestTestfile.cmake
-    - DMD_TESTSUITE_MAKE_ARGS='-j16 GDB_FLAGS=OFF' ctest -V -R "dmd-testsuite" || true
-    # Run defaultlib unittests (non-debug only for now, excl. hanging core.thread.fiber)
+    - DMD_TESTSUITE_MAKE_ARGS='-j32 GDB_FLAGS=OFF' ctest -V -R "dmd-testsuite" || true
+    # Run defaultlib unittests (excl. hanging core.thread.fiber)
     # & druntime stand-alone tests, ignore the errors
-    - ctest -j16 --output-on-failure -E "dmd-testsuite|lit-tests|ldc2-unittest|-debug(-shared)?$|^core.thread.fiber($|-)" || true
+    - ctest -j32 --output-on-failure -E "dmd-testsuite|lit-tests|ldc2-unittest|^core.thread.fiber($|-)" || true
     # Install LDC & make portable
     - ninja install > /dev/null
     - cd ..

--- a/tests/codegen/sret_thunk_gh3329.d
+++ b/tests/codegen/sret_thunk_gh3329.d
@@ -1,0 +1,42 @@
+// RUN: %ldc -run %s
+
+extern(C) int printf(const(char)* format, ...);
+
+struct NoPOD
+{
+    size_t x;
+    ~this() {}
+}
+
+interface I
+{
+    NoPOD doIt(size_t arg);
+}
+
+__gshared C c;
+
+class C : I
+{
+    this()
+    {
+        c = this;
+        printf("c: %p\n", c);
+    }
+
+    NoPOD doIt(size_t arg)
+    {
+        printf("doIt this: %p; arg: %p\n", this, arg);
+        assert(this == c);
+        assert(arg == 0x2A);
+        return NoPOD(arg << 4);
+    }
+}
+
+void main()
+{
+    I i = new C;
+    printf("i: %p\n", i);
+    NoPOD r = i.doIt(0x2A);
+    printf("&r: %p\n", &r);
+    assert(r.x == 0x2A0);
+}

